File: 41.cpp
#include <math.h>
#include <vector>
using namespace std;

class Solution {
private:
  bool inBounds(int a, vector<int> &nums) {
    if (a < nums.size() && a > -1) {
      return true;
    }
    return false;
  }

public:
  int firstMissingPositive(vector<int> &nums) {
    for (int &i : nums) {
      if (i < 0) {
        i = 0;
      }
    }
    for (int i = 0; i < nums.size(); ++i) {
      int idx = abs(nums[i]) - 1;
      if (inBounds(idx, nums)) {
        int val = nums[idx];
        if (val == 0) {
          nums[idx] = -(nums.size() + 1);
        } else {
          nums[idx] = -abs(nums[idx]);
        }
      }
    }
    for (int cand = 1; cand < nums.size() + 1; ++cand) {
      if (nums[cand - 1] >= 0) {
        return cand;
      }
    }
    return nums.size() + 1;
  }
};

int main() { Solution s; }

File: 41.py
class Solution:
    def firstMissingPositive(self, nums: list[int]) -> int:
        for i in range(len(nums)):
            if nums[i] < 0:
                nums[i] = 0
        for i in range(len(nums)):
            idx = abs(nums[i]) - 1
            if idx > -1 and idx < len(nums):
                if nums[idx] != 0:
                    nums[idx] = -abs(nums[idx])
                else:
                    nums[idx] = -abs(len(nums) + 1)
        for i in range(1, len(nums) + 1, 1):
            if nums[i - 1] >= 0:
                return i
        return len(nums) + 1


assert Solution().firstMissingPositive([1, 2, 0]) == 3
assert Solution().firstMissingPositive([3, 4, -1, 1]) == 2
assert Solution().firstMissingPositive([7, 8, 9, 11, 12]) == 1
